crates/rattler_installs_packages/src/wheel.rs

```diff
@@ -0,0 +1,878 @@
+/// for more information.
+pub struct Wheel {
+    name: WheelName,
+    archive: Mutex<ZipArchive<Box<dyn ReadAndSeek + Send>>>,
+}
+
+impl Wheel {
+    /// Open a wheel by reading a file on disk.
+    pub fn from_path(path: &Path) -> miette::Result<Self> {
+        let file_name = path
+            .file_name()
+            .and_then(OsStr::to_str)
+            .ok_or_else(|| miette::miette!("path does not contain a filename"))?;
+        let wheel_name = WheelName::from_str(file_name).into_diagnostic()?;
+        let file = File::open(path).into_diagnostic()?;
+        Self::new(wheel_name, Box::new(file))
+    }
+}
+
+impl Artifact for Wheel {
+    type Name = WheelName;
+
+    fn new(name: Self::Name, bytes: Box<dyn ReadAndSeek + Send>) -> miette::Result<Self> {
+        Ok(Self {
+            name,
+            archive: Mutex::new(ZipArchive::new(bytes).into_diagnostic()?),
+        })
+    }
+
+    fn name(&self) -> &Self::Name {
+        &self.name
+    }
+}
+
+impl Wheel {
+    /// A wheel file always contains a special directory that contains the metadata of the package.
+    /// This function returns the name of that directory.
+    fn find_special_wheel_dir<'a>(
+        top_level_names: impl IntoIterator<Item = &'a str>,
+        name: &PackageName,
+        version: &Version,
+        suffix: &str,
+    ) -> Result<Option<&'a str>, WheelVitalsError> {
+        // Find all directories that end in the suffix
+        let mut candidates = top_level_names.into_iter().filter(|dir_name| {
+            let Some(candidate) = dir_name.strip_suffix(suffix) else {
+                return false;
+            };
+            let Some((candidate_name, candidate_version)) = candidate.rsplit_once('-') else {
+                return false;
+            };
+
+            let Ok(candidate_name) = PackageName::from_str(candidate_name) else {
+                return false;
+            };
+            let Ok(candidate_version) = Version::from_str(candidate_version) else {
+                return false;
+            };
+
+            &candidate_name == name && &candidate_version == version
+        });
+
+        // Get the first candidate
+        let candidate = match candidates.next() {
+            Some(candidate) => candidate,
+            None => return Ok(None),
+        };
+
+        // Error out if there are multiple directories
+        if candidates.next().is_some() {
+            return Err(WheelVitalsError::MultipleSpecialDirs(suffix.to_owned()));
+        }
+
+        Ok(Some(candidate))
+    }
+
+    async fn get_lazy_vitals(
+        name: &WheelName,
+        stream: &mut AsyncHttpRangeReader,
+    ) -> Result<(Vec<u8>, WheelCoreMetadata), WheelVitalsError> {
+        // Make sure we have the back part of the stream.
+        // Because the zip index is at the back
+        stream
+            .prefetch(stream.len().saturating_sub(16384)..stream.len())
```

Comment:
What is `16384`?

Reply:
Ah good question! Thats my best guess at the size of the central archive directory (the index). Sometimes its more, which then causes another range requests but most of the times it less.